src/Microsoft.Extensions.DependencyInjection/ServiceProvider.cs

```diff
@@ -188,6 +191,8 @@ public void Dispose()
 
         internal object CaptureDisposable(object service)
         {
+            _captureDisposableCallback?.Invoke(service);
```

Comment:
Inspecting the expressions definitely produces cleaner product code. You just don't want to write more complicated tests. The test is super fragile to begin with because it's testing an implementation detail that isn't at all observable given using black box testing. Maybe the perf difference is observable, but we haven't even established that yet.